### PR TITLE
[clang][analyzer] Bring checker 'alpha.unix.cstring.NotNullTerminated' out of alpha

### PR DESCRIPTION
**Pull request by balazske**
I have tested it on the usual amount of C and C++ projects, no results were found. This checker is only for very rare cases, I do not know if it is really useful but it is there and can be moved out of alpha. There is no test when only this checker is enabled (and not full unix.cstring), but I have checked and it works too.

### CODE REVIEW

**Review by NagyDonat — COMMENTED** [review]
Overall I support moving this checker out of alpha, because its narrow functionality indeed seems to be stable enough for general use.

However, eventually it would be good to extend this checker to get a "real" NonNullTerminated checker that can deduce that certain "normal" strings don't have a proper null terminator.

File: clang/docs/analyzer/checkers.rst [hunk 1]
```diff
@@ -1899,6 +1899,29 @@ Check the size argument passed into C string functions for common erroneous patt
...
+
+Only very fundamental cases are detected where the passed memory block is
+absolutely different from a null-terminated string. This checker does not
+find if a memory buffer is passed where the terminating zero character
+is missing.
+
+.. code-block:: c
+
+ void test1() {
+   int l = strlen((char *)&test); // warn
```
Comment (steakhal) [inline clang/docs/analyzer/checkers.rst:1917]: I think you wanted to take the address of test1.

File: clang/test/Analysis/string.cpp [hunk 2]
```diff
@@ -53,3 +53,7 @@ struct TestNotNullTerm {
     strlen((char *)&x); // expected-warning{{Argument to string length function is not a null-terminated string}}
   }
 };
+
+void test_notcstring_tempobject() {
+  strlen((char[]){'a', 0}); // expected-warning{{Argument to string length function is a C++ temp object of type char[2], which is not a null-terminated string}}
+}
```
[the remark this answers is not in the document]
Comment (balazske) [reply clang/test/Analysis/string.cpp:59]: This is taken from a FIXME in line 991 of the checker code. This exact case is really a false positive, a warning is needed only if a write happens into such a region. For this exact case (temporary object that behaves like a C string) the checker can be improved. But such a change would affect other checkers in the 'cstring' group.

**Comment by balazske** [comment]
I have added a FIXME to the place where no warning should appear. Additionally a new test file is added where only this checker is turned on (not the full `cstring`).

**Review by NagyDonat — APPROVED** [review]
LGTM, thanks for the updates.

**Review by steakhal — APPROVED** [review]

**Pull request merged**